File: python/src/control_box.py
import math
from time import sleep
from enum import Enum
from presence_detector import Pir
from led_strip import LedStrip


STEPS = 100
TIME_STEP = 0.010
MAX_DT = 100


class Mode(str, Enum):
    DETECTOR  = 'DETECTOR'
    DIMMER    = 'DIMMER'
    BREATHING = 'BREATH'
    BLINK     = 'BLINK'


class ControlBox:
    def __init__(self, detectPin):
        self.led = LedStrip()
        self.pir = Pir(detectPin, self.led)
        self.mode = Mode.DETECTOR
        self.brightness = 0
        self.toggle = False
        self.breathing_increasing = []
        self.breathing_decreasing = []
        self.breathing_index = 0
        self.__prepare_breathing()

    def __prepare_breathing(self):
        for i in range(STEPS):
            brightness = MAX_DT * (1 - math.exp(-4 * i / STEPS))
            self.breathing_increasing.append(brightness)

        for i in range(STEPS):
            self.breathing_decreasing.append(self.breathing_increasing[STEPS - 1 - i])

    def processDimmerCommand(self, newBrightness):
        if newBrightness < 0:
            newBrightness = 0
        elif newBrightness > 100:
            newBrightness = 100

        self.brightness = newBrightness

        if self.mode is Mode.DIMMER:
            self.led.set_pwm(newBrightness)

    def set_mode(self, new_mode: str) -> None:
        if new_mode.lower() == Mode.DETECTOR.lower():
            self.mode = Mode.DETECTOR
            self.pir.reset()
            return

        if new_mode.lower() == Mode.DIMMER.lower():
            self.mode = Mode.DIMMER
            self.led.set_pwm(self.brightness)
            return

        if new_mode.lower() == Mode.BLINK.lower():
            self.mode = Mode.BLINK
            self.led.set_pwm(self.brightness)
            return

        if new_mode.lower() == Mode.BREATHING.lower():
            self.mode = Mode.BREATHING
            self.led.set_pwm(0)
            self.toggle = False
            self.breathing_index = 0
            return

    def run(self):
        if self.mode == Mode.DETECTOR:
            self.pir.poll()
            sleep(0.1)
        elif self.mode == Mode.BLINK:
            if self.toggle:
                self.led.turn_off()
            else:
                self.led.set_pwm(self.brightness)
            self.toggle = not self.toggle
            sleep(1)
        elif self.mode == Mode.BREATHING:
            if self.toggle:
                self.led.set_pwm(self.breathing_decreasing[self.breathing_index])
            else:
                self.led.set_pwm(self.breathing_increasing[self.breathing_index])
            self.breathing_index += 1
            if self.breathing_index >= STEPS:
                self.breathing_index = 0
                self.toggle = not self.toggle
            sleep(TIME_STEP)
        else:
            sleep(0.25)

File: cpp/src/control_box.cpp
#include <unistd.h>
#include <cmath>
#include <unordered_map>
#include "control_box.hpp"


#define DETECT_POLLING_PERIOD_US 100000
#define DIM_POLLING_PERIOD_US    250000
#define BREATH_POLLING_PERIOD_US 10000
#define BLINK_POLLING_PERIOD_US  1000000


ControlBox::ControlBox(SysfsPwm& ledStrip, int detectPin)
: PresenceDetector(ledStrip, detectPin)
{
    brightnessDutyCycle_ = 0;
    stepIndex_ = 0;
    toggle_ = false;
    currentMode_ = Mode::DETECT;
    prepareBreath();
}

void ControlBox::doMode(void)
{
    switch (currentMode_)
    {
    case Mode::DETECT:
        poll();
        usleep(DETECT_POLLING_PERIOD_US);
        break;

    case Mode::DIM:
        usleep(DIM_POLLING_PERIOD_US);
        break;

    case Mode::BREATH:
        doBreath();
        usleep(BREATH_POLLING_PERIOD_US);
        break;

    case Mode::BLINK:
        doBlink();
        usleep(BLINK_POLLING_PERIOD_US);
        break;

    default:
        break;
    }
}

void ControlBox::setMode(const std::string& modeStr)
{
    ControlBox::Mode newMode = stringToMode(modeStr);
    currentMode_ = newMode;
    switch (newMode)
    {
    case Mode::DETECT:
        reset();
        break;

    case Mode::DIM:
        ledStrip_.setDutyCycle(brightnessDutyCycle_);
        break;

    case Mode::BREATH:
        ledStrip_.turnOff();
        toggle_ = false;
        stepIndex_ = 0;
        break;

    case Mode::BLINK:
        ledStrip_.setDutyCycle(brightnessDutyCycle_);
        toggle_ = false;
        break;

    default:
        currentMode_ = Mode::DETECT;
        break;
    }
}

void ControlBox::setBrightness(int newBrightness)
{
    if (newBrightness > 100) {
        brightnessDutyCycle_ = 100;
        return;
    }

    if (newBrightness < 0) {
        brightnessDutyCycle_ = 0;
        return;
    }

    brightnessDutyCycle_ = newBrightness;
}

void ControlBox::doBlink(void)
{
    if (toggle_) {
        ledStrip_.setDutyCycle(brightnessDutyCycle_);
    }
    else {
        ledStrip_.turnOff();
    }
    toggle_ = !toggle_;
}

void ControlBox::doBreath(void)
{
    int newDt = toggle_ ? decreasingDt_[stepIndex_] : increasingDt_[stepIndex_];
    ledStrip_.setDutyCycle(newDt);
    stepIndex_++;
    if (stepIndex_ >= STEPS)
    {
        toggle_ = !toggle_;
        stepIndex_ = 0;
    }
}

void ControlBox::prepareBreath(void)
{
    for (int i = 0; i < STEPS; i++)
    {
        float brightness = (float)MAX_DT * (1.f - exp(-4.f * (float)i / (float)STEPS));
        increasingDt_[i] = (int)brightness;
        decreasingDt_[STEPS - i - 1] = (int)brightness;
    }
}

ControlBox::Mode ControlBox::stringToMode(const std::string& modeStr)
{
    static std::unordered_map<std::string, Mode> map = {
        {"Detector", Mode::DETECT},
        {"Dimmer", Mode::DIM},
        {"Blink", Mode::BLINK},
        {"Breath", Mode::BREATH}
   };

   auto it = map.find(modeStr);
   if (it == map.end()) {
        throw std::invalid_argument("Invalid mode string: " + modeStr);
   }

   return it->second;
}